File: 202309/725.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func splitListToParts(head *ListNode, k int) []*ListNode {
	n := 0
	p := head
	for p != nil {
		n++
		p = p.Next
	}
	ave := n / k
	more := n % k
	p = head
	var ans []*ListNode
	var prev *ListNode = nil

	for i := 0; i < k; i++ {
		q := p
		ans = append(ans, q)
		for j := 0; j < ave; j++ {
			prev = p
			p = p.Next
		}
		if more > 0 {
			more--
			prev = p
			p = p.Next
		}
		if prev != nil {
			prev.Next = nil
		}
	}
	return ans
}

File: 202409/725.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def splitListToParts(
        self, head: Optional[ListNode], k: int
    ) -> List[Optional[ListNode]]:
        n = 0
        p = head
        while p:
            n += 1
            p = p.next

        a, b = divmod(n, k)
        p = head
        ans = []
        for _ in range(k):
            ans.append(p)
            q = p
            for _ in range(a):
                q = p
                p = p.next
            if b > 0:
                q = p
                p = p.next
                b -= 1
            if q is not None:
                q.next = None

        return ans
